src/main/java/com/microsoft/graph/tasks/LargeFileUploadTask.java

```diff
@@ -0,0 +1,323 @@
+                               @Nonnull Parsable uploadSession,
+                               @Nonnull InputStream uploadStream,
+                               long streamSize,
+                               long maxSliceSize,
+                               @Nonnull ParsableFactory<T> factory) throws IllegalAccessException, IOException, InvocationTargetException, NoSuchMethodException {
+        Objects.requireNonNull(uploadSession);
+        Objects.requireNonNull(uploadStream);
+        Objects.requireNonNull(factory);
+        if(uploadStream.available() <=0) {
+            throw new IllegalArgumentException("Must provide a stream that is not empty.");
+        }
+        this.uploadSession = extractSessionFromParsable(uploadSession);
+        this.requestAdapter = Objects.isNull(requestAdapter) ? initializeAdapter(this.uploadSession.getUploadUrl()):requestAdapter;
+        this.totalUploadLength = streamSize;
+        this.rangesRemaining = getRangesRemaining(this.uploadSession);
+        this.uploadStream = uploadStream;
+        this.maxSliceSize = maxSliceSize;
+        this.factory = factory;
+    }
+    /**
+     * Perform the upload task.
+     * @return An UploadResult model containing the information from the server resulting from the upload request.
+     */
+    @Nonnull
+    public CompletableFuture<UploadResult<T>> uploadAsync() {
+        return this.uploadAsync(3, null);
+    }
+    /**
+     * Perform the upload task.
+     * @param maxTries Number of times to retry the task before giving up.
+     * @param progress IProgress interface describing how to report progress.
+     * @return An UploadResult model containing the information from the server resulting from the upload request.
+     */
+    @Nonnull
+    public CompletableFuture<UploadResult<T>> uploadAsync(int maxTries, @Nullable IProgressCallback progress) {
+        int uploadTries = 0;
+        ArrayList<Throwable> exceptionsList = new ArrayList<>();
+        while (uploadTries < maxTries) {
+            try {
+                List<UploadSliceRequestBuilder<T>> uploadSliceRequestBuilders = getUploadSliceRequests();
+                for (UploadSliceRequestBuilder<T> request : uploadSliceRequestBuilders) {
+                    UploadResult<T> result;
+                    result = uploadSliceAsync(request, exceptionsList);
+                    amountUploaded += request.getRangeLength();
+                    if(progress != null) {
+                        progress.report(amountUploaded, this.totalUploadLength);
+                    }
+                    if (result.isUploadSuccessful()) {
+                        return CompletableFuture.completedFuture(result);
+                    }
+                }
+                updateSessionStatusAsync().get();
+                uploadTries += 1;
+                if (uploadTries < maxTries) {
+                    TimeUnit.SECONDS.sleep((long) 2 * uploadTries * uploadTries);
+                }
```

Comment:
The retry in this scenario is not from a per-request basis (which indeed the retry handler will handle) but from the perspective of the entire upload.
An example is when a client timeout may occur when uploading a slice, and the client will then have no idea if the chunk was uploaded successfully. In this scenario, it would then use the upload session url to ask the server what chunks haven't been received yet (therefore confirming if the timed-out request was successful or not by calling`updateSessionStatusAsync().get();`) and then use that information to upload the remaining chunks to prevent a possible conflict error from the API end due to a replay of the same chunk.